crypto/src/mls/conversation/commit.rs

```diff
@@ -92,13 +116,24 @@ impl CentralContext {
     /// If the conversation can't be found, an error will be returned. Other errors are originating
     /// from OpenMls and the KeyStore
     #[cfg_attr(test, crate::idempotent)]
-    pub async fn update_keying_material(&self, id: &ConversationId) -> CryptoResult<MlsCommitBundle> {
-        let client = self.mls_client().await?;
+    pub async fn update_keying_material(&self, id: &ConversationId) -> Result<MlsCommitBundle> {
+        let client = self
+            .mls_client()
+            .await
+            .map_err(RecursiveError::root("getting mls client"))?;
```

Comment:
This repeated in quite a few places. Are we ever getting calling `mls_client()` with another intention than "getting mls client"?

Reply:
Same question also applies for `mls_provider()`

Reply:
No, but I don't see a better way to structure this.

The `.mls_client()` method (and `mls_provider()` as well) returns a root error: `core_crypto::Error`, implemented in `crypto/src/error/mod.rs`. But we call that method from all over the place, and we need to create all kinds of different errors from that. In this case what the function actually returns is a conversation error: `core_crypto::mls::conversation::Error`, defined in `crypto/src/mls/conversation/error.rs`. 

You can see for yourself what's going on here: `RecursiveError::root` returns a function which wraps the root error and adds the context line, and then the `?` operator uses the `From<RecursiveError>` implementation (from the `#[from]` annotation) to actually convert the error. But the point of the context field is to distinguish the cases. Quite likely, this "getting mls client" error is not the top-level error. It's common to get a MLS client and provider very near each other, and both operations will end up being mapped with `RecursiveError::root`. Having this context allows us to distinguish what in particular is being returned there.

Maybe you're thinking that the answer would be to have the `mls_client` method do its own wrapping: instead of returning `Result<Client, core_crypto::Error>`, it might return `Result<Client, E> where RecursiveError: Into<E>`. That would allow us to inline the context at least, avoid explicitly specifying that we're getting the mls client each time we get the mls client. Unfortunately, there are problems with that approach:

1. It breaks the pattern that each group of module always returns the internal error type for that group of modules. Now we're returning _something_ based on this internal error type, but we can't say exactly what until we know the context of our caller.
2. It forces us to wrap everything everywhere in `RecursiveError`. That makes it more difficult, not less, to determine what the internal error type is.
3. It breaks the `?` operator. The reason for this is that this defines the returned type in terms of `Into`. But the `?` operator also applies an `Into` conversion. So desugaring everything looks like:
    ```rust
    let client = match self.mls_client() {
      Ok(client) => client,
      Err(e) => return e.into(),
    };
    ```
    But Rust can't infer what type `e` should have there, because any number of types might potentially satisfy both `into` conditions. 

Maybe I'm missing something. But I do think that this somewhat-repetitive pattern is the best option we have for now.

(But it's worth mentioning that avoiding this kind of mess is a big driver of my desire to do a big refactoring that we can talk about next month! I don't have the domain knowledge yet to come into that week with a plan for what precisely we should do. But my hope is by the end of the week we will have designed a structure that lets avoid `RecursiveError` entirely.)